pkg/controllers/vrep/replication_reconciler.go

```diff
@@ -238,33 +239,74 @@ func (r *ReplicationReconciler) checkSandboxExists() error {
 	return nil
 }
 
-// choose the source host and target host
-// (first host where db is up in the specified cluster)
-func (r *ReplicationReconciler) determineSourceAndTargetHosts() (err error) {
-	// assume source could be read-only, no subcluster constraints
-	upPodIP, ok := r.SourcePFacts.FindFirstUpPodIP(true, "")
-	if !ok {
-		err = fmt.Errorf("cannot find any up hosts in source database cluster")
+// filters services based on svcName if not empty and main/sandbox cluster
+// picks first service that matches. if svcName is empty this will be the first service assigned to the database
+// if svcName is non-empty it will be the service matching that name
+// returns a valid hostname for the service or an error if a valid matching service couldn't be found
+func (r *ReplicationReconciler) getServiceHostname(ctx context.Context, vdb *vapi.VerticaDB,
+	sandboxName, svcName string) (ip string, err error) {
+	finder := iter.MakeSubclusterFinder(r.VRec.Client, vdb)
+	var stss *appsv1.StatefulSetList
+	stss, err = finder.FindStatefulSets(ctx, iter.FindInVdb, sandboxName)
+	if err != nil {
+		return "", err
+	}
+
+	for i := range stss.Items {
```

Comment:
An easier way would be:
- If svcName is non-empty: to iterate over vdb.spec.subclusters, get the servicename(`names.GenExtSvcName(vdb, sc)`). check if it is equal svcName, check if the sc sandboxName is equal to sandboxName.
- If svcName is empty: iterate over vdb.spec.subclusters, pick the 1st primary subcluster's service. main cluster primaries have `type = primary` while sandbox primaries have `type = sandboxPrimary`.

To know if a sc belongs to a sandbox you need to check the status. There are already functions like `vdb.GenSubclusterSandboxStatusMap()` or `vdb.GetSandboxStatus()` that can help.